contracts/utils/Strings.sol

```diff
@@ -64,4 +64,36 @@ library Strings {
         require(value == 0, "Strings: hex length insufficient");
         return string(buffer);
     }
+
+    /**
+     * @dev Converts each character ('a' to 'z') of a `string` to it's upper-case form ('A' to 'Z').
+     * Requires parameter string `value` to be comprised purely of basic-ASCII characters.
+     */
+    function normalizeUpperASCII(string calldata value) internal pure returns (string memory) {
+        bytes memory buffer = bytes(value);
+        for (uint256 i = 0; i < buffer.length; i++) {
+            require(buffer[i] & 0x80 == 0, "Basic ASCII string required."); // 0x80-bit on is non-ASCII.
+            if (buffer[i] >= "a" && buffer[i] <= "z") { // only work on lower-case characters.
+                buffer[i] &= 0x1F; // mask-off the lower-case bit.
+                assert(buffer[i] >= "A" && buffer[i] <= "Z"); // assert invariance in change to uppercase.
```

Comment:
This is guaranteed by math ... why check it? it brings nothing and adds bytecode.

Reply:
It's to protect against (potential) bugs in the math. Afaik, this is how assert() is meant to be used - at the end to validate invariance in code / results. This helps guarantee the function always returns a normalized string. I think since the scope of this function is to normalize strings, there's some value in having this guard in place. I don't know how much bytecode it adds, I assumed it wouldn't be too much.

Reply:
I do agree with your philosophy behind the usage of assert, I just don't think it applies here. IMO the math is simple enough that we can be certain this will never be triggered... And if there was still a doubt, I would rather use formal verification than shipping extra bytecode.

Reply:
I'm not opposed to removing the assert() if that's the consensus.